torch/csrc/aot_inductor/aten_shim_common.cpp

```diff
@@ -0,0 +1,492 @@
+//    See the comments in aten_tensor_to_aot_tensor and
+//    aot_tensor_to_aten_tensor for more details.
+// 2. sizes and strides are not dynamically allocated and managed here. They are
+//    either coming from input/output aten tensors, or are codegen-ed as
+//    constant arrays by AOTInductor. In both cases, sizes and strides are
+//    guaranteed to be live during the model's execution. In a few cases,
+//    AOTInductor may need to codegen extra arguments to for an op's output
+//    tensor's sizes and strides, e.g. at::convolution.
+// 3. As for AOTInductorTensor objects, they are dynamically allocated and
+//    managed by std::unique_ptr. They can also be statically planned and
+//    allocated later.
+struct AOTInductorTensor {
+  void* data_ptr;
+  void* deleter;
+  int64_t ndim;
+  const int64_t* sizes;
+  const int64_t* strides;
+  int64_t storage_offset;
+  int8_t device_type;
+  int8_t device_index;
+  int8_t dtype;
+  int padding : 24;
+
+  AOTInductorTensor(
+      void* data_ptr,
+      void* deleter,
+      int64_t ndim,
+      const int64_t* sizes,
+      const int64_t* strides,
+      int64_t storage_offset,
+      int8_t device_type,
+      int8_t device_index,
+      int8_t dtype)
+      : data_ptr(data_ptr),
+        deleter(deleter),
+        ndim(ndim),
+        sizes(sizes),
+        strides(strides),
+        storage_offset(storage_offset),
+        device_type(device_type),
+        device_index(device_index),
+        dtype(dtype),
+        padding(0) {}
+
+#ifndef NDEBUG
+  void print() {
+    std::cout << "data_ptr=" << data_ptr << ", ";
+    std::cout << "deleter=" << deleter << ", ";
+    std::cout << "ndim=" << ndim << ", ";
+    std::cout << "sizes=" << sizes << ", ";
+    std::cout << "strides=" << strides << ", ";
+    std::cout << "storage_offset=" << storage_offset << ", ";
+    std::cout << "device_type=" << int32_t(device_type) << ", ";
+    std::cout << "device_index=" << int32_t(device_index) << ", ";
+    std::cout << "dtype=" << int32_t(dtype) << std::endl;
+  }
+#endif
+};
+
+size_t get_dtype_bytes(AOTInductorScalarType dtype) {
+  switch (dtype) {
+    case kAOTInductorBool:
+    case kAOTInductorByte:
+    case kAOTInductorChar:
+    case kAOTInductorQInt8:
+    case kAOTInductorQUInt8:
+      return 1;
+    case kAOTInductorShort:
+    case kAOTInductorHalf:
+    case kAOTInductorBFloat16:
+      return 2;
+    case kAOTInductorInt:
+    case kAOTInductorFloat:
+    case kAOTInductorComplexHalf:
+    case kAOTInductorQInt32:
+      return 4;
+    case kAOTInductorLong:
+    case kAOTInductorDouble:
+    case kAOTInductorComplexFloat:
+      return 8;
+    case kAOTInductorComplexDouble:
+      return 16;
+    default:
+      AOT_INDUCTOR_CHECK(false, "Unkown AOTInductorScalarType")
+      return 0;
+  }
+}
+
+// TODO: this probably can also be statically planned to avoid any dynamic
+// allocation
+thread_local std::vector<std::unique_ptr<AOTInductorTensor>> tensor_manager_;
```

Comment:
Instead of shoving this in a thread local, it seems like it would more explicit to represent the runtime state as an explicit struct and explicitly pass it around. This doesn't seem like a big burden to the model.so, which is all code-genned anyway. Otherwise, you need to clearly document the threading requirements (specifically, model.so is never allowed to migrate to another thread; you are not allowed to run multiple runtimes on the same thread, etc)